### PR TITLE
feat (invoices): add mutation that retry all failed and pending invoices

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

Batch retry all failed (pending) invoices

## Description

From the UI we want to allow users to retry all failed/pending invoices.

### CODE REVIEW

File: spec/graphql/mutations/invoices/retry_all_payments_spec.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+          current_organization: organization,
+          current_user: user,
+          query: mutation,
+          variables: {
+            input: {},
+          },
+        )
+
+        data = result['data']['retryAllPayments']
+        invoice_ids = data['collection'].map { |value| value['id']}
```
Comment (jdenquin) [inline spec/graphql/mutations/invoices/retry_all_payments_spec.rb:64]: Layout/SpaceInsideBlockBraces: Space missing inside }.

File: app/jobs/invoices/payments/retry_all_job.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Invoices
+  module Payments
+    class RetryAllJob < ApplicationJob
+      queue_as 'invoices'
+
+      def perform(organization_id:, invoice_ids:)
+        result = Invoices::Payments::RetryBatchService.new(organization_id: organization_id).retry_all(invoice_ids)
```
Comment (jdenquin) [inline app/jobs/invoices/payments/retry_all_job.rb:9]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: app/services/invoices/payments/retry_batch_service.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  module Payments
+    class RetryBatchService < BaseService
+      def initialize(organization_id:)
+        @organization_id = organization_id
+
+        super
+      end
+
+      def call
+        Invoices::Payments::RetryAllJob.perform_later(organization_id: organization_id, invoice_ids: invoices.ids)
```
Comment (jdenquin) [inline app/services/invoices/payments/retry_batch_service.rb:13]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: app/services/invoices/payments/retry_batch_service.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+        result.invoices = invoices
+
+        result
+      end
+
+      def retry_all(invoice_ids)
+        processed_invoices = []
+        Invoice.where(id: invoice_ids).each do |invoice|
+          result = Invoices::Payments::RetryService.new(invoice: invoice).call
```
Comment (jdenquin) [inline app/services/invoices/payments/retry_batch_service.rb:23]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: spec/graphql/mutations/invoices/retry_all_payments_spec.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Mutations::Invoices::RetryAllPayments, type: :graphql do
+  let(:membership) { create(:membership) }
+  let(:organization) { membership.organization }
+  let(:user) { membership.user }
+  let(:gocardless_payment_provider) { create(:gocardless_provider, organization: organization) }
```
Comment (jdenquin) [inline spec/graphql/mutations/invoices/retry_all_payments_spec.rb:9]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: spec/graphql/mutations/invoices/retry_all_payments_spec.rb [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  let(:customer_first) { create(:customer, organization: organization, payment_provider: 'gocardless') }
```
Comment (jdenquin) [inline spec/graphql/mutations/invoices/retry_all_payments_spec.rb:10]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: spec/graphql/mutations/invoices/retry_all_payments_spec.rb [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  let(:gocardless_customer_first) { create(:gocardless_customer, customer: customer_first) }
+  let(:customer_second) { create(:customer, organization: organization, payment_provider: 'gocardless') }
```
Comment (jdenquin) [inline spec/graphql/mutations/invoices/retry_all_payments_spec.rb:12]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: spec/services/invoices/payments/retry_batch_service_spec.rb [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+      end.to have_enqueued_job(Invoices::Payments::RetryAllJob)
+    end
+  end
+
+  describe '#retry_all' do
+    let(:invoice_ids) { [invoice_first.id, invoice_second.id] }
+    let(:invoice_first) do
+      create(
+        :invoice,
+        customer: customer,
```
Comment (jdenquin) [inline spec/services/invoices/payments/retry_batch_service_spec.rb:24]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: spec/services/invoices/payments/retry_batch_service_spec.rb [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        status: 'finalized',
+        payment_status: 'failed',
+        ready_for_payment_processing: true,
+      )
+    end
+    let(:invoice_second) do
+      create(
+        :invoice,
+        customer: customer,
```
Comment (jdenquin) [inline spec/services/invoices/payments/retry_batch_service_spec.rb:33]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: spec/services/invoices/payments/retry_batch_service_spec.rb [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        status: 'finalized',
+        payment_status: 'failed',
+        ready_for_payment_processing: true,
+      )
+    end
+    let(:invoice_third) do
+      create(
+        :invoice,
+        customer: customer,
```
Comment (jdenquin) [inline spec/services/invoices/payments/retry_batch_service_spec.rb:42]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: spec/services/invoices/payments/retry_batch_service_spec.rb [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+      expect(processed_ids).to include(invoice_first.id)
+      expect(processed_ids).to include(invoice_second.id)
+      expect(processed_ids).not_to include(invoice_third.id)
+    end
+
+    context 'when inner service passes error result' do
+      let(:invoice_second) do
+        create(
+          :invoice,
+          customer: customer,
```
Comment (jdenquin) [inline spec/services/invoices/payments/retry_batch_service_spec.rb:71]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: app/services/invoices/payments/retry_batch_service.rb [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+      def call
+        Invoices::Payments::RetryAllJob.perform_later(organization_id:, invoice_ids: invoices.ids)
+
+        result.invoices = invoices
+
+        result
+      end
+
+      def retry_all(invoice_ids)
```
Comment (vincent-pochet) [inline app/services/invoices/payments/retry_batch_service.rb:20]: Just an idea: what about having this one name `call` and the first one `call_later`?

Just wondering if we should keep `call` name for the method that is performing the business logic.

That's maybe something we could discuss in the back team.

**Review by vincent-pochet — APPROVED** [review]
Looks great!

**Pull request merged**